core/trino-main/src/main/java/io/trino/operator/scalar/ArrayVectorFunctions.java

```diff
@@ -0,0 +1,83 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.operator.scalar;
+
+import io.trino.spi.block.Block;
+import io.trino.spi.function.Description;
+import io.trino.spi.function.ScalarFunction;
+import io.trino.spi.function.SqlType;
+
+import static io.trino.spi.StandardErrorCode.INVALID_FUNCTION_ARGUMENT;
+import static io.trino.spi.type.RealType.REAL;
+import static io.trino.spi.type.StandardTypes.DOUBLE;
+import static io.trino.util.Failures.checkCondition;
+
+public final class ArrayVectorFunctions
+{
+    private ArrayVectorFunctions() {}
+
+    @Description("Calculates the euclidean distance between two vectors")
+    @ScalarFunction
+    @SqlType(DOUBLE)
+    public static double euclideanDistance(@SqlType("array(real)") Block first, @SqlType("array(real)") Block second)
```

Comment:
Why does this take `array(real)` and return `double`? It should probably return `real`, and there should be a variant that takes `array(double)`.

Reply:
Changing the return type to `real` leads to ±inifinty even when it's no problem as `double` return type. Do you still think it should be separated?

Reply:
What kind of input causes that? I’m surprised the result would be much larger than then magnitude of the input vectors. Are you doing the intermediate computations with doubles or with floats?

Reply:
The example is `euclidean_distance(ARRAY[REAL '3.4028235e+38f'], ARRAY[REAL '-3.4028235e+38f'])` and it was a computation with floats. 

The result with double is `6.805646932770577E38`. I suppose that is outside of valid `real` type either?

Reply:
That seems like an extreme case. It’s like saying “why does adding two integers overflow” (the sum of two integers returns integer). If you want to avoid overflow for numbers that are close to the limit, you need to first promote them to a higher precision type.

Reply:
Updated. Now I think having only `array(double)` is sufficient because the engine internally handles the type difference.